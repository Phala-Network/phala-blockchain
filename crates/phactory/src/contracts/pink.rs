use crate::contracts;
use crate::system::{TransactionError, TransactionResult};
use anyhow::{anyhow, Result};
use parity_scale_codec::{Decode, Encode};
use phala_mq::traits::MessageChannel;
use phala_mq::{ContractGroupId, ContractId, MessageOrigin};
use runtime::{AccountId, BlockNumber};

#[derive(Debug, Encode, Decode)]
pub enum Command {
    InkMessage { nonce: Vec<u8>, message: Vec<u8> },
}

#[derive(Debug, Encode, Decode)]
pub enum Query {
    InkMessage(Vec<u8>),
}

#[derive(Debug, Encode, Decode)]
pub enum Response {
    InkMessageReturn(Vec<u8>),
}

#[derive(Debug, Encode, Decode)]
pub enum QueryError {
    BadOrigin,
    RuntimeError(String),
}

pub struct Pink {
    instance: pink::Contract,
    group: ContractGroupId,
}

impl Pink {
    pub fn instantiate(
        group: ContractGroupId,
        storage: &mut pink::Storage,
        origin: AccountId,
        wasm_bin: Vec<u8>,
        input_data: Vec<u8>,
        salt: Vec<u8>,
        block_number: BlockNumber,
        now: u64,
    ) -> Result<Self> {
        let instance = pink::Contract::new(
            storage,
            origin.clone(),
            wasm_bin,
            input_data,
            salt,
            block_number,
            now,
        )
        .map_err(|err| anyhow!("Instantiate contract failed: {:?} origin={:?}", err, origin,))?;
        Ok(Self { group, instance })
    }
}

impl contracts::NativeContract for Pink {
    type Cmd = Command;

    type QReq = Query;

    type QResp = Result<Response, QueryError>;

    fn id(&self) -> ContractId {
        let inner: &[u8; 32] = self.instance.address.as_ref();
        inner.into()
    }

    fn handle_query(
        &mut self,
        origin: Option<&AccountId>,
        req: Query,
        context: &mut contracts::QueryContext,
    ) -> Result<Response, QueryError> {
        let origin = origin.ok_or(QueryError::BadOrigin)?;
        match req {
            Query::InkMessage(input_data) => {
                let storage = group_storage(&mut context.contract_groups, &self.group)
                    .expect("Pink group should always exists!");

                let (ret, _messages) = self
                    .instance
                    .bare_call(
                        storage,
                        origin.clone(),
                        input_data,
                        true,
                        context.block_number,
                        context.now_ms,
                    )
                    .map_err(|err| {
                        log::error!("Pink [{:?}] query exec error: {:?}", self.id(), err);
                        QueryError::RuntimeError(format!("Call contract method failed: {:?}", err))
                    })?;
                return Ok(Response::InkMessageReturn(ret));
            }
        }
    }

    fn handle_command(
        &mut self,
        origin: MessageOrigin,
        cmd: Command,
        context: &mut contracts::NativeContext,
    ) -> TransactionResult {
        match cmd {
            Command::InkMessage { nonce: _, message } => {
                let origin: runtime::AccountId = match origin {
                    MessageOrigin::AccountId(origin) => origin.0.into(),
                    _ => return Err(TransactionError::BadOrigin),
                };

                let storage = group_storage(&mut context.contract_groups, &self.group)
                    .expect("Pink group should always exists!");

                let (ret, messages) = self
                    .instance
                    .bare_call(
                        storage,
                        origin.clone(),
                        message,
                        false,
                        context.block.block_number,
                        context.block.now_ms,
                    )
                    .map_err(|err| {
                        log::error!("Pink [{:?}] command exec error: {:?}", self.id(), err);
                        TransactionError::Other(format!("Call contract method failed: {:?}", err))
                    })?;

                // TODO.kevin: store the output to some where.
                let _ = ret;

                for message in messages.messages {
                    context.mq.push_data(message.payload, message.topic);
                }

                for message in messages.osp_messages {
                    context
                        .secret_mq
                        .bind_remote_key(message.remote_pubkey.as_ref())
                        .push_data(message.payload, message.topic);
                }
            }
        }
        Ok(())
    }
}

fn group_storage<'a>(
    groups: &'a mut group::GroupKeeper,
    group_id: &ContractGroupId,
) -> Result<&'a mut pink::Storage> {
    groups
        .get_group_storage_mut(group_id)
        .ok_or(anyhow!("Contract group {:?} not found! qed!", group_id))
}

pub mod group {
    use super::Pink;

    use crate::contracts::support::NativeContract as _;
    use anyhow::Result;
    use phala_mq::{ContractGroupId, ContractId};
    use pink::types::AccountId;
    use runtime::BlockNumber;
    use std::collections::{BTreeMap, BTreeSet};

    #[derive(Default)]
    pub struct GroupKeeper {
        groups: BTreeMap<ContractGroupId, Group>,
    }

    #[derive(Default)]
    pub struct Group {
        storage: pink::Storage,
        contracts: BTreeSet<ContractId>,
    }

    impl GroupKeeper {
        pub fn instantiate_contract(
            &mut self,
            group_id: ContractGroupId,
            origin: AccountId,
            wasm_bin: Vec<u8>,
            input_data: Vec<u8>,
            salt: Vec<u8>,
            block_number: BlockNumber,
            now: u64,
        ) -> Result<Pink> {
            let group = self.groups.entry(group_id.clone()).or_default();
            let pink = Pink::instantiate(
                group_id,
                &mut group.storage,
                origin,
                wasm_bin,
                input_data,
                salt,
                block_number,
                now,
            )?;
            group.contracts.insert(pink.id().clone());
            Ok(pink)
        }

        pub fn get_group_storage_mut(
            &mut self,
            group_id: &ContractGroupId,
        ) -> Option<&mut pink::Storage> {
            Some(&mut self.groups.get_mut(group_id)?.storage)
        }
    }
}

pub mod messaging {
    use parity_scale_codec::{Decode, Encode};
    use phala_crypto::sr25519::Sr25519SecretKey;
    use phala_mq::{bind_topic, ContractGroupId};
    use phala_types::WorkerPublicKey;
    use pink::types::AccountId;

    pub use phala_types::messaging::{ContractInfo, WorkerPinkReport};

    bind_topic!(WorkerPinkRequest, b"phala/pink/worker/request");
    #[derive(Encode, Decode, Debug)]
    pub enum WorkerPinkRequest {
        Instantiate {
            group_id: ContractGroupId,
            worker: WorkerPublicKey,
            nonce: Vec<u8>,
            owner: AccountId,
            wasm_bin: Vec<u8>,
            input_data: Vec<u8>,
            salt: Vec<u8>,
            key: Sr25519SecretKey,
        },
    }

    bind_topic!(GKPinkRequest, b"phala/pink/gk/request");
    #[derive(Encode, Decode, Debug)]
    pub enum GKPinkRequest {
        Instantiate {
            group_id: Option<ContractGroupId>, // None for create a new one
            worker: WorkerPublicKey, // TODO: None for choosing one by GK or by the group_id?
            wasm_bin: Vec<u8>,
            input_data: Vec<u8>,
        },
    }
}

#[test]
fn test_make_pink_request() {
    use crate::secret_channel::Payload;
    use hex_literal::hex;

    let request = messaging::WorkerPinkRequest::Instantiate {
        group_id: Default::default(),
        worker: phala_types::WorkerPublicKey(hex!(
            "3a3d45dc55b57bf542f4c6ff41af080ec675317f4ed50ae1d2713bf9f892692d"
        )),
        nonce: vec![],
        owner: hex!("3a3d45dc55b57bf542f4c6ff41af080ec675317f4ed50ae1d2713bf9f892692d").into(),
        wasm_bin: include_bytes!("fixtures/flip.contract").to_vec(),
        input_data: hex!("9bae9d5e01").to_vec(),
        salt: vec![],
        key: [0; 64],
    };
    let message = Payload::Plain(request);
    println!("message: {}", hex::encode(message.encode()));
}
