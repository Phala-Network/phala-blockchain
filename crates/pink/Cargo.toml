[package]
name = "pink"
version = "0.1.0"
edition = "2018"

[dependencies]
sha2 = "0.9.6"
log = "0.4.14"
pallet-contracts = { path = "../../substrate/frame/contracts" }
pallet-contracts-primitives = { path = "../../substrate/frame/contracts/common" }
pallet-contracts-proc-macro = { path = "../../substrate/frame/contracts/proc-macro" }
pallet-balances = { path = "../../substrate/frame/balances" }
pallet-timestamp = { path = "../../substrate/frame/timestamp" }
pallet-randomness-collective-flip = { path = "../../substrate/frame/randomness-collective-flip" }

frame-support = { path = "../../substrate/frame/support" }
frame-system = { path = "../../substrate/frame/system" }

sp-core = { path = "../../substrate/primitives/core" }
sp-runtime = { path = "../../substrate/primitives/runtime" }
sp-io = { path = "../../substrate/primitives/io" }
sp-std = { path = "../../substrate/primitives/std" }
sp-sandbox = { path = "../../substrate/primitives/sandbox" }
sp-state-machine = { path = "../../substrate/primitives/state-machine" }
sp-externalities = { path = "../../substrate/primitives/externalities" }

scale = { package = "parity-scale-codec", version = "2.0.0", default-features = false, features = ["derive"] }
parity-wasm = { version = "0.41.0", default-features = false }
pwasm-utils = { version = "0.16", default-features = false }
wasmi-validation = { version = "0.3.0", default-features = false }

paste = "1.0"
pretty_assertions = "0.7.2"
wat = "1.0"
call-trace = "*"
impl-serde = "0.3.1"
hex = "0.4.3"
serde = { version = "1.0.101", features = ["derive"] }
serde_json = "1.0.67"

pink-extension = { path = "pink-extension" }

[dev-dependencies]
insta = "1.7.2"
hex-literal = "0.3.3"