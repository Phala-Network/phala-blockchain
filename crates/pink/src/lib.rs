extern crate alloc;

mod contract;
mod export_fixtures;
pub mod local_cache;

pub mod runtime;
pub mod storage;

pub mod types;

pub use contract::{Contract, ContractFile, Storage, transpose_contract_result};
pub use export_fixtures::load_test_wasm;
