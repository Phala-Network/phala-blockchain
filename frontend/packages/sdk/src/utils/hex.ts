import randomBytes from 'randombytes'
import { type HexString } from '../types'

export const randomHex = (size = 12): HexString => randomBytes(size).toString('hex')
