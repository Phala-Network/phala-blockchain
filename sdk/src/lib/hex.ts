import {randomBytes} from 'crypto-browserify'

export const randomHex = (size = 12): string =>
  randomBytes(size).toString('hex')
