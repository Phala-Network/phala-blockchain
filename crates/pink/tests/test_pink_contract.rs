use frame_support::assert_ok;
use hex_literal::hex;
use pink::Contract;
use sp_runtime::AccountId32;

pub const ALICE: AccountId32 = AccountId32::new([1u8; 32]);

#[test]
fn test_ink_flip() {
    let mut storage = Contract::new_storage();
    let mut contract = Contract::new_with_selector(
        &mut storage,
        ALICE.clone(),
        include_bytes!("./fixtures/flip/flip.wasm").to_vec(),
        hex!("9bae9d5e"), // init_value
        true,
        vec![],
    )
    .unwrap();

    let result: bool = contract
        .call_with_selector(
            &mut storage,
            ALICE.clone(),
            hex!("2f865bd9"), // get
            (),
            false,
        )
        .unwrap();

    assert_eq!(result, true); // Should equal to the init value

    let _: () = contract
        .call_with_selector(
            &mut storage,
            ALICE.clone(),
            hex!("633aa551"), // flip
            (),
            false,
        )
        .unwrap();

    let result: bool = contract
        .call_with_selector(
            &mut storage,
            ALICE.clone(),
            hex!("2f865bd9"), // get
            (),
            false,
        )
        .unwrap();

    assert_eq!(result, false); // Should be flipped

    let result: (u32, u128) = contract
        .call_with_selector(
            &mut storage,
            ALICE.clone(),
            hex!("f7dff04c"), // echo
            (42u32, 24u128),
            false,
        )
        .unwrap();
    assert_eq!(result, (42, 24));
}

#[test]
fn test_load_contract_file() {
    assert_ok!(pink::ContractFile::load(include_bytes!(
        "./fixtures/flip/flip.contract"
    )));
}

#[test]
fn test_ink_cross_contract_instanciate() {
    let mut storage = Contract::new_storage();
    let _flip = Contract::new_with_selector(
        &mut storage,
        ALICE.clone(),
        include_bytes!("./fixtures/flip/flip.wasm").to_vec(),
        hex!("9bae9d5e"), // init_value
        true,
        vec![],
    )
    .unwrap();

    let mut contract = Contract::new_with_selector(
        &mut storage,
        ALICE.clone(),
        include_bytes!("./fixtures/cross/cross.wasm").to_vec(),
        hex!("9bae9d5e"),
        (),
        vec![],
    )
    .unwrap();

    let result: bool = contract
        .call_with_selector(
            &mut storage,
            ALICE.clone(),
            hex!("c3220014"), // get
            (),
            false,
        )
        .unwrap();

    insta::assert_debug_snapshot!(result);
}

#[test]
fn test_mq_egress() {
    let mut storage = Contract::new_storage();
    let mut contract = Contract::new_with_selector(
        &mut storage,
        ALICE.clone(),
        include_bytes!("./fixtures/mqproxy/mqproxy.wasm").to_vec(),
        hex!("ed4b9d1b"), // init_value
        (),
        vec![],
    )
    .unwrap();

    let _: () = contract
        .call_with_selector(
            &mut storage,
            ALICE.clone(),
            hex!("6495da7f"), // push_message
            (b"\x42\x42".to_vec(), b"\x24\x24".to_vec()),
            false,
        )
        .unwrap();

    let _: () = contract
        .call_with_selector(
            &mut storage,
            ALICE.clone(),
            hex!("d09d68e0"), // push_osp_message
            (
                b"\x42\x42".to_vec(),
                b"\x24\x24".to_vec(),
                Some([0u8; 32]),
            ),
            false,
        )
        .unwrap();

    let messages = pink::runtime::take_mq_egress();
    insta::assert_debug_snapshot!(messages);
}
