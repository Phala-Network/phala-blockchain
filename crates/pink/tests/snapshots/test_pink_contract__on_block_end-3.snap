---
source: crates/pink/tests/test_pink_contract.rs
expression: effects

---
ExecSideEffects {
    pink_events: [
        (
            b81c488581e2d18bdb886027d0b848b76f2376723c9b2fadf5ccc5b23a2ea038 (5GE751Bk...),
            Message(
                Message {
                    payload: [
                        102,
                        111,
                        111,
                    ],
                    topic: [
                        47,
                        98,
                        97,
                        114,
                    ],
                },
            ),
        ),
    ],
    instantiated: [],
}
